Add process_sleepspindle_parameters written by @park-minchul

park-minchul opened the pull request.

park-minchul committed:
Add process_sleepspindle_parameters written by @mcp0228

park-minchul committed:
Update process_sleepspindle_parameters

park-minchul commented:
Process has been updated with the following amendments:
1. Sampling frequency is no longer hardcoded but calculated by the number of time samples-1 divided by the duration
2. Counting the number of zeros and sign changes for calculation of frequency has been removed. This is now replaced by counting the number of spindle peaks.
3. Clicking online tutorial now takes the user to the GitHub README.md file written by the author to describe how the process works. 

I've decided not to add a channel selection parameter in the options because spindles are by definition an EEG phenomenon therefore it wouldn't make sense for a user to select EOG or EMG for this process.

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+    sProcess.Category    = 'Custom';
+    sProcess.SubGroup    = 'Custom Processes';
+    sProcess.Index       = 1000;
+    sProcess.Description = 'https://github.com/mcp0228/Brainstorm-Custom-Processes/tree/main/process_sleepspindle_parameters#process-sleep-spindle-parameters';
```
Link broken, it points to the previous username

park-minchul replied:
new link provided - which will take the user to the README page. 
README file uploaded here. 
[README.md](https://github.com/brainstorm-tools/bst-users/files/13290098/README.md)

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+    sProcess.nMinFiles   = 1;
+
+    % Description of the process
+    sProcess.options.info.Comment = ['Takes in X number of sleep spindles and calculates the following parameters.<BR><BR>'...
```
Indicate that each input data file correspond to one sleep spindle

park-minchul replied:
wording updated

    % Description of the process
    sProcess.options.info.Comment = ['Takes in X number of sleep spindles (each input data file = one spindle) and calculates the following parameters.<BR><BR>'...
                                     '1. Spindle duration in seconds.<BR>'...

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+                                     'Nat Methods . 2014 Apr;11(4):385-92. doi: 10.1038/nmeth.2855."<BR><BR>'...
+                                     'Notes<BR>'...
+                                     'A) This process assumes that the data was already filtered between 11-16 Hz.<BR>'...
+                                     'B) This process will generate 4 files - duration, frequency, amplitude and symmetry.<BR>'...
```
Indicate that the output is 4 matrix files containing the features extracted from all the input files

park-minchul replied:
wording updated to 

```
'A) This process assumes that the data was already filtered between 11-16 Hz.<BR>'...
'B) This process will generate 4 matrix files containing duration, frequency, amplitude and symmetry features from all input files.<BR>'...
```

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+    % Initialize returned list of files
+    OutputFiles = {};
+    Nepochs     = length(sInputs); % Numer of inputs or epochs
+    EpochN      = linspace (1,(Nepochs),(Nepochs)); % Changes the x-axis from time (s) to epoch number n. 
```
Simpler:
`EpochN = 1 : Nepochs;`

park-minchul replied:
updated as suggested.

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+    % ===== PROCESS =====
+    % This is where the actual process of data manipulation and calculation takes place.
+       DataN        = Data*-1; % The negative version of the data
+       Data(end,:)  = (randi(100,1,Ntime))*10^-6; % End row which is BDF/status channel is a constant number so replaced this with randi to remove issues during findpeaks 
+       DataN(end,:) = (randi(100,1,Ntime))*10^-6;
```
Instead of setting the last channel to noise, it will be better to run the FOR loop only for EEG channels, and check that there are not flat.


```
% Get channel file for each input file
sChannel = bst_get('ChannelForStudy', [sInputs(iNepochs).iStudy]);
% Load channel information
ChannelMat = in_bst_channel(ChannelFile, 'Channel');
% Indices for EEG channels only
eegIxs = strcmpi({ChannelMat.Channel.Type}, 'EEG');

% Current FOR loop per channel
for eegIx = 1 : length(eegIxs)
    i = eegIxs(eegIx);
    Data_Chan = Data(i,:); % Extracts individual channel info from the data matrix   
    % Check for flat channel
    if sum((Data_Chan - mean(Data_Chan)).^2) == 0
        continue;
    end
    % Compute sleep spindles ...
end
```

rcassani replied:
Note that the detection of flat channels could be optimized. I added in the FOR loop to make clearer

park-minchul replied:
Thanks for the input @rcassani.
I've made some adjustments but keep getting the following error message which I cannot resolve. What are your thoughts? 
![image](https://github.com/brainstorm-tools/bst-users/assets/134780775/aac44457-a01f-4419-a9c9-6133f1febe72)

```
    % ===== LOAD THE DATA =====
    for iNepochs = 1 : Nepochs
       DataMat   = in_bst(sInputs(iNepochs).FileName, [], 0);
       epochSize = size(DataMat.F);  
       Ntime     = epochSize(2); % Number of time samples changes depending on sleep spindle
       Data      = DataMat.F; % Actual data containing EEG, EOG and EMG data
       Time      = DataMat.Time(end) - DataMat.Time(1); % Time information of the recording
       Fs        = 1 ./ (DataMat.Time(2) - DataMat.Time(1)); % Calculation of Sampling frequency

       % Get channel file for each input file
       sChannel = bst_get('ChannelForStudy', [sInputs(iNepochs).iStudy]);
       % Load channel information
       ChannelMat = in_bst_channel(sChannel, 'Channel');
       % Indices for EEG channels only
       eegIxs = strcmpi({ChannelMat.Channel.Type}, 'EEG');

    % ===== PROCESS =====
    % This is where the actual process of data manipulation and calculation takes place.
       DataN        = Data*-1; % The negative version of the data
       %Data(end,:)  = (randi(100,1,Ntime))*10^-6; % End row which is BDF/status channel is a constant number so replaced this with randi to remove issues during findpeaks 
       %DataN(end,:) = (randi(100,1,Ntime))*10^-6;

            %for i = 1 : Nchannels
            for eegIx = 1 : length (eegIxs)
                i = eegIxs(eegIx);
                Data_Chan = Data(i,:); % Extracts individual channel info from the data matrix 
                if sum((Data_Chan - mean(Data_Chan)).^2) == 0 % Check for flat channel
                 continue;
                end
% compute spindles...
```

rcassani replied:
Solved at [3baad30](https://github.com/brainstorm-tools/bst-users/pull/9/commits/3baad304b4620d7a3ee434890d6fa7d87cfe38f8)

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+       DataN_Chan   = zeros(1,Ntime); % Row vector with Ntime X zeros
+
+            for i = 1 : Nchannels
+                Data_Chan(1,:)     = Data(i,:); % Extracts individual channel info from the data matrix
+                DataN_Chan(1,:)    = DataN(i,:); % Extracts individual channel info from the negative data matrix
+                [PosP, LocP]       = findpeaks(Data_Chan,Fs); % Find all the positive peaks and their locations in seconds
+                [NegP, LocN]       = findpeaks(DataN_Chan,Fs); % Find all the negative peaks and their locations in seconds
```
This creates a copy of the data for one channel. While it may have a small impact in memory use and performance, it is not a good practice. It could be done as:

`[PosP, LocP] = findpeaks(Data(i, :), Fs);`

park-minchul replied:
amended as per your suggestion!

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+       SS_Amp             = PeakPo + PeakNe; % Final calculation of spindle max peak-peak amplitude
+       SS_Sym             = (TimePP./SS_Dur)*100; % Final calculation of spindle symmetry
+
+    % Spindle duration = the number of time samples-1 divided by the sampling frequency of 1024 
+    % Spinde frequency = the reciprocal of (the number of peaks/duration)
+    % Spinde amplitude = the sum of max positive peak and negative peak
+    % Spindle symmetry = the percentage of (the time location of max positive peak/spindle duration) 
```
Each of these comments would be better placed in the line before the feature computation

park-minchul replied:
updated as appropriate

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+       SS_Amp             = PeakPo + PeakNe; % Final calculation of spindle max peak-peak amplitude
+       SS_Sym             = (TimePP./SS_Dur)*100; % Final calculation of spindle symmetry
+
+    % Spindle duration = the number of time samples-1 divided by the sampling frequency of 1024 
```
In comment, Fs is not fixed to 1024Hz

park-minchul replied:
removed the comment of 1024 Hz

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+    db_add_data(iStudy, OutputFiles{1}, DataMat);
+end
+
+% ===== Resources Used =====
+% https://au.mathworks.com/matlabcentral/answers/240706-how-do-i-make-a-code-that-will-count-the-number-of-times-the-sign-changes-succesively-through-an-arr?s_tid=ta_ans_results
+% MATLAB code for counting the number of incidences where a sign change occurs
+% https://au.mathworks.com/matlabcentral/answers/214893-counting-the-number-of-non-zero-or-zero-elements-in-a-specific-direction-of-an-array?s_tid=ta_ans_results
+% MATLAB code for counting the number of zeros in an EEG epoch across channels (i.e. rows)
+% https://au.mathworks.com/matlabcentral/answers/2028920-nested-for-loop-adding-elements-to-a-matrix
+% MATLAC code for adding elements in a nested for loop in a multidimensional array
+% https://neuroimage.usc.edu/forums/t/custom-process-for-individual-epochs-with-different-time-samples/42260
+% Brainstorm user forum about making a custom process for epochs with variable time samples
```
Move to the README.md file

park-minchul replied:
Done as requested! :)

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+% MATLAC code for adding elements in a nested for loop in a multidimensional array
+% https://neuroimage.usc.edu/forums/t/custom-process-for-individual-epochs-with-different-time-samples/42260
+% Brainstorm user forum about making a custom process for epochs with variable time samples
+
+% ===== Redundant Code For Counting the No. of Zeros and Sign Changes =====
+% NoOfSC = zeros (Nchannels,Nepochs); % Number of sign changes
+% NoOfZe = zeros (Nchannels,Nepochs); % Number of zeros     
+% Data_Pos = Data>0; % Converts Data so that numbers>0 are 1 & <0 are 0
+% ChanZero = zeros(1,Ntime); % Row vector with Ntime X zeros
+% ChanZero(1,:) = Data_Pos(i,:); % Extracts individual channel info from the data>0 matrix
+% SChanges = xor(ChanZero(1:end-1),ChanZero(2:end)); % Count the number of incidences where a sign change occurs
+% NoOfSC(i,iNepochs) = sum(SChanges); % Number of sign changes per channel
+% NoOfZe(:,iNepochs) = sum(Data==0,2); % Count of no. of zeros 
+% Count = (NoOfZe + NoOfSC)/2; % Final count of no. of zeros & no. of sign changes
```
To be removed?

park-minchul replied:
removed as appropriate

rcassani commented on processes/process_sleepspindle_parameters/process_sleepspindle_parameters:
```diff
@@ -0,0 +1,202 @@
+       epochSize = size(DataMat.F);  
+       Ntime     = epochSize(2); % Number of time samples changes depending on sleep spindle
+       Data      = DataMat.F; % Actual data containing EEG, EOG and EMG data
+       Time      = DataMat.Time(end); % Time information of the recording
+       Fs        = (Ntime-1)/Time; % Calculation of Sampling frequency
```
Safer to compute as below. E.g, in the case if negative values in the Time vector

```
Time = DataMat.Time(end) - DataMat.Time(1);
Fs   = 1 ./ (DataMat.Time(2) - DataMat.Time(1));
```

park-minchul replied:
updated as appropriate

rcassani reviewed (changes requested):
The code looks good, there some things to address, see the review comments.

Also, a README.md file is missing, to provide a brief description of the code.
Something like this one: https://github.com/brainstorm-tools/bst-users/blob/master/processes/process_average_subsets/README.md

It can contain link to your README with more details

park-minchul committed:
Update process_sleepspindle_parameters

park-minchul committed:
Update process_sleepspindle_parameters

rcassani committed:
Add `.m` extension to process file

rcassani committed:
Process only EEG channels

rcassani committed:
Save spindle parameters as matrix files (instead of recording files)

rcassani commented:
Hi @park-minchul, I made two changes to the process:
1. It now runs only for EEG channels rather than for all channels in the recording
2. Results are saved as `matrix` (instead of recordings `data` files)

It seems, a useful part would be to have a part of the code to verify:
* That all recordings come from the same Study, therefore they share the same channel file 
  (is this necessarily true?)
  
* That all recordings have the same good channels, e.g, to avoid computing sleep spindle parameters in channels that were previously labelled as bad

park-minchul commented:
Hi @rcassani Thanks for your help in updating this process! 
I just tried this on a few of my recordings and it seems that if a recording has other channels such as EOG and EMG then you get an error message like the below; 

> ** Error: [process_sleepspindle_parameters]  Custom Processes > Sleep spindle parameters
** Line 92: Unable to perform assignment because the size of the left side is 1-by-1 and the size of the right side is 1-by-0.
** 
** Call stack:
** >process_sleepspindle_parameters.m>Run at 92
** >process_sleepspindle_parameters.m at 11
** >bst_process.m>Run at 236
** >bst_process.m at 38
** >panel_process1.m>RunProcess at 138
** >panel_process1.m at 26
** >gui_brainstorm.m>CreateWindow/ProcessRun_Callback at 777
** >bst_call.m at 28
** >gui_brainstorm.m>@(h,ev)bst_call(@ProcessRun_Callback) at 298
** 
** 
***************************************************************************

The replies to your questions are as below; 
1. That all recordings come from the same Study, therefore they share the same channel file (is this necessarily true?)
- yes; this would be a useful feature. Depending on the type of protocol and the study, recordings may not share the same channel file. 

2. That all recordings have the same good channels, e.g, to avoid computing sleep spindle parameters in channels that were previously labelled as bad
- yes; this would be a useful feature indeed. 
- another feature that would be useful for this process would be to exclude "BAD" trials. 

I see that you've also added a feature to check for FLAT channels - thank you for this too - this is very useful. 

One question from me, what was your reason for results being saved as MATRIX vs RECORDING?

park-minchul committed:
Update process_sleepspindle_parameters.m

Added two new parameters; spindle RMS amplitude as well as spindle activity (peak to peak amplitude*duration).

park-minchul committed:
Update process_sleepspindle_parameters.m

changed sProcess.options.info.comment to update added features.

rcassani committed:
Bugfi: Find EEG indices

rcassani committed:
Error at building Comment

rcassani committed:
Correct indexing for EEG channels

rcassani committed:
Compute SS_Rms only for EEG channels

rcassani commented:
Error fixed.

1. For this point, the files will be grouped by Study, then the spindle parameters will be computed. This is to say, there will a set of matrices for set of recording files that belong to the same Study. 

2. An error message (and process stop) will occur if the Good/Bad channels is different for the files in the same Study. BAD trials are already excluded from processing. 

3. The spindle parameters are saved as matrix files to avoid having the Kernel projection to these files. Plotting of the features is still available.

park-minchul commented:
@rcassani thank you so so much!!! 👍

park-minchul commented:
Error found @ [16d0ca8](https://github.com/brainstorm-tools/bst-users/pull/9/commits/16d0ca82c6870014aeb16408d9827973f61ffe3d) the results produce matrix file where the RMS values for each spindle is identical across all channels.

park-minchul committed:
Update process_sleepspindle_parameters.m

error at [16d0ca8](https://github.com/brainstorm-tools/bst-users/pull/9/commits/16d0ca82c6870014aeb16408d9827973f61ffe3d) updated so that all columns are processed for RMS

rcassani commented:
@park-minchul, It seems this PR is ready to be merge 🎉 
Please verify that it is working as expected and let me know to merge it to the principal branch

park-minchul commented:
Dear @rcassani thank you for your help 🙏 
I can confirm that this is working fine both as an individual process and as part of a script!! 👌
